src/main/kotlin/org/rust/lang/core/psi/ext/RsAbstractable.kt

```diff
@@ -6,7 +6,54 @@
 package org.rust.lang.core.psi.ext
 
 import com.intellij.psi.PsiNameIdentifierOwner
+import org.rust.lang.core.psi.*
 
 interface RsAbstractable : RsNamedElement, PsiNameIdentifierOwner {
     val isAbstract: Boolean
 }
+
+sealed class RsAbstractableOwner {
+    object Free : RsAbstractableOwner()
+    object Foreign : RsAbstractableOwner()
+    class Trait(val trait: RsTraitItem) : RsAbstractableOwner()
+    class Impl(val impl: RsImplItem, val isInherent: Boolean) : RsAbstractableOwner()
+
+    val isInherentImpl: Boolean get() = this is Impl && isInherent
+    val isTraitImpl: Boolean get() = this is Impl && !isInherent
+    val isImplOrTrait: Boolean get() = this is Impl || this is Trait
+}
+
+val RsAbstractable.owner: RsAbstractableOwner
+    get() {
+        val stubOnlyParent = when (this) {
```

Comment:
I could not find a nice way to do this for all 3 different types. Hope this is OK.